### ISSUE TITLE
CI in OQS: guidelines for responsible use

### ISSUE THREAD

**Issue by bhess** [open-quantum-safe/tsc#5]
This issue is to suggest setting guidelines in OQS for resource-responsible use of CI.

> OQS so far already did much more than "typical projects" in this regards/not blindly running CI in the various projects -- but there's still quite some way to go: What about OQS "institutes" this? Ask contributors to use "[skip ci]" as often as possible, do (CI) spot checks before doing "heavy lifting", possibly move more CI to the weekly tests, reduce the repetition of the profiling runs, etc?

_Originally posted by @baentsch in https://github.com/open-quantum-safe/oqs-provider/issues/376#issuecomment-2006940977_

I agree with the above and think some guidelines would help to avoid redundant CI runs, without hurting the overall test quality. This is not necessarily motivated by a lack of available resources (the LF indicated that more Github Actions resources can be made available to the projects, which I think is a good thing), but rather by the desire to make most efficient use of the available resources.

**Comment by SWilson4** [comment]
I agree that this is a good idea. One "obvious" way to make CI more efficient would be to eliminate redundant runs (e.g., the same workflow being triggered by both "push" and "pull_request" on some PRs). I haven't yet found a clean way to do this, but perhaps somebody with more GitHub-specific expertise might.

IMO this is a great place to contribute for those who want to help out OQS / PQCA but don't feel comfortable working with cryptographic code.

**Comment by dstebila** [comment]
> I agree that this is a good idea. One "obvious" way to make CI more efficient would be to eliminate redundant runs (e.g., the same workflow being triggered by both "push" and "pull_request" on some PRs). I haven't yet found a clean way to do this, but perhaps somebody with more GitHub-specific expertise might.

@ryjones might have experience with this.

**Comment by SWilson4** [comment]
> @SWilson4 do you mean on PR and merge, or the same job runs on both PR and the push of the PR to a branch?

The latter. For example, see the "Linux and macOS tests" here: https://github.com/open-quantum-safe/liboqs/pull/1725/checks. These tests (and all others triggered to run on both `push` and `pull_request` events run twice on every push to an existing PR (at least for my PRs and presumably non-fork PRs).

**Comment by ryjones** [comment]
@SWilson4 [this stack overflow answer](https://stackoverflow.com/a/61565445) shows the way.

**Comment by ryjones** [comment]
[take a look here](https://github.com/hyperledger/besu/blob/5ea9cd64c2f4886d3efa7bad3539a5adfa0e3eea/.github/workflows/integration-tests.yml#L1-L12)

**Comment by dstebila** [comment]
The outcome of today's OQS status call was that it would be good to have a document setting some guidelines / documenting best practices, saved as a Markdown file in the TSC repository. @bhess and @thb-sb to kick it off.

**Comment by baentsch** [comment]
Thanks very much for documenting decisions such as this, @dstebila .

May I augment with the suggestion to link the file thus generated from a (new) "getting started contributing to OQS" write-up at www.openquantumsafe.org? Maybe off/in/at https://openquantumsafe.org/about/#getting-started (?)

**Comment by SWilson4** [comment]
In the absence of codified guidelines, here are some principles I tried to follow in https://github.com/open-quantum-safe/liboqs/pull/1909. I would suggest that we work toward following these in other projects as well.

1. Cancel in-progress jobs on a double-push. This can be achieved using [concurrency groups](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/control-the-concurrency-of-workflows-and-jobs#example-only-cancel-in-progress-jobs-or-runs-for-the-current-workflow).
2. Only run resource-heavy tests if basic checks pass first.
3. Run only a minimal set of tests automatically on push to non-main branches. 
4. As a follow-up to the above, give developers the ability to trigger heavier tests manually—ideally both by pushing a button in the GitHub UI (using [workflow dispatches](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_dispatch)) and by including a keyword in the commit message.

I found [reusable workflows](https://docs.github.com/en/actions/sharing-automations/reusing-workflows#creating-a-reusable-workflow) to be very helpful for implementing all of these. These allow us to create "caller" workflows for each condition (push, pull request, release, schedule, etc) which trigger a subset of separate "callee" workflows (macOS tests, Linux tests, downstream tests, extended tests, etc). With this pattern we can get very fine-grained control over which tests run and when.

**Comment by bhess** [comment]
# CI Usage Guidelines
The OQS project aims to reduce redundant CI runs while maintaining good test quality, promoting efficient use of computational resources, minimizing environmental impact, and ensuring developers receive timely feedback from CI results.

## Contributor Guidelines

To help achieve this goal, contributors are asked to follow these practices:

- Use `[skip ci]` in commit messages for changes that do not affect program logic and do not require testing (e.g., documentation updates, formatting).
- Prefer manual workflow triggers via GitHub Actions when selective testing is sufficient.
- Cancel superseded workflows manually if a newer commit renders the current CI run obsolete.
- Use `[full tests]` in commit messages only when the change requires testing across all supported platforms.
- Use `[extended tests]` only when the change affects features covered by extended tests (e.g., algorithm behavior changes requiring extended KAT or constant-time tests).
- For PRs requiring workflow approval, the approver should perform a basic sanity check before approving the CI run.

## Project-Level CI Optimization Goals

To maximize CI efficiency and ensure effective use of resources, the OQS CI setup should aim to:

- Ensure basic workflows complete within a few minutes.
- Gate resource-heavy tests behind successful completion of basic checks.
- Limit automatic test runs on non-main branches to a minimal set.
- Cancel in-progress jobs on double-pushes.
- Enable manual triggers for heavy or extended test workflows.
- Ensure regular CI runs in PRs do not exceed X hours in total runtime.

If exceeded, consider:

- Adding parallel jobs
- Moving long-running jobs to weekly workflows

Optimize Weekly Workflows: If total runtime exceeds acceptable limits (i.e.., X time), reduce the frequency of these tests to monthly or pre-release runs.

**Comment by bhess** [comment]
# CI Usage Guidelines
The OQS project aims to reduce redundant CI runs while maintaining good test quality, promoting efficient use of computational resources, minimizing environmental impact, and ensuring developers receive timely feedback from CI results.

## Contributor Guidelines

To help achieve this goal, contributors are asked to follow these practices:

- Use `[skip ci]` in commit messages for changes that do not affect program logic and do not require testing (e.g., documentation updates, formatting).
  - Example: `git commit -s -m "chore: correcting a typo in the documentation [skip ci]"`
- Prefer manual workflow triggers via GitHub Actions when selective testing is sufficient.
  - Workflows that have a "workflow_dispatch" trigger can be triggered manually. To trigger them, go to "Actions", select the workflow in the list on the left, then click on "Run workflow" and select the branch or the commit to be tested.
<img width="864" height="508" alt="Image" src="https://github.com/user-attachments/assets/c723823a-b948-4dbf-a7ec-55860a314718" />
- Cancel superseded workflows manually if a newer commit renders the current CI run obsolete.
  - To cancel a workflow, go to "Actions", select the workflow to be cancelled, and click "Cancel workflow"
  <img width="864" height="508" alt="Image" src="https://github.com/user-attachments/assets/102acc72-2b4d-473e-8939-7d75b9ff6c43" />

- Use `[full tests]` in commit messages only when the change requires testing across all supported platforms.
  - Example: `git commit -s -m "<your commit message> [full tests]"`
- Use `[extended tests]` only when the change affects features covered by extended tests (e.g., algorithm behavior changes requiring extended KAT or constant-time tests).
  - Example: `git commit -s -m "<your commit message> [extended tests]"`

## Reviewer Guidelines

- For pull requests requiring workflow approval, the approver should do a basic sanity check before approving the CI run.

## Project-Level CI Optimization Goals

To maximize CI efficiency and ensure effective use of resources, the OQS CI setup should aim to:

- Ensure basic workflows complete within 2 minutes.
- Gate resource-heavy tests behind successful completion of basic checks.
- Ensure that push tests without any tags (i.e., without `[full tests]` or `[extended tests]`) complete within 5 minutes.
- Ensure that PR tests without any tags (i.e., without `[full tests]` or `[extended tests]`) complete within 15 minutes.
- Ensure that full tests (tagged with `[full tests]`) complete within 30 minutes.
- Ensure that extended tests (tagged with `[extended tests]`) complete within 1 hour.
- Limit automatic test runs on non-main branches to a minimal set.
- Cancel in-progress jobs on double-pushes.
- Enable manual triggers for heavy or extended test workflows.
- Ensure regular CI runs in PRs do not exceed 1 hour until completion.

If the limits are exceeded, consider:

- Adding parallel jobs.
- Moving longer-running jobs to weekly workflows (up to 2 hours).
- Moving jobs running longer than 2 hours to tests only triggered manually.
- All workflows, including those only triggered manually, should be run before a release.

**Comment by bhess** [comment]
Thank you for your feedback, @baentsch.
I’ve updated the guidelines in the message above to include time limits and added screenshots showing how to manually trigger workflows and cancel them.

> Then putting this all into an OQS-global Wiki with links pointing to it from each subproject's CONTRIBUTING.md and we can close this issue as far as I'm concerned.

Do we currently have an OQS-global Wiki? If not, one option could be to add this content to the liboqs wiki and link to it from each subproject’s CONTRIBUTING.md.
